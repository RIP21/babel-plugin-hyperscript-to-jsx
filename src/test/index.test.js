"use strict";

const pluginTester = require("babel-plugin-tester");
const hyperToJsxTransform = require("../index.js");
const prettier = require("prettier");
const { check, complexExample } = require("./cases");

const hAndhxImports = `import h from 'h';
import hx from 'hx';`;

const tests = [
  {
    title: "Fake all cases",
    code: check
  },
  { title: "Complex real", code: complexExample },
  {
    title: "Import default added",
    code: `
    import { Component } from 'react';
    import h from 'hyper';
    `
  },
  {
    title: "Import default is not added cause no hyperscript",
    code: `
    import { Component } from 'react';
    `
  },
  {
    title: "If default import is there, keep it with no affect",
    code: `
    import React, { Component } from 'react';
    import h from 'hyper';
    `
  },
  {
    title: "Should handle ArrowFunctions shortcut return",
    code: `${hAndhxImports}
   const StatelessComponent = props => h("h1");`
  },
  {
    title: "Should handle ArrowFunctions with return",
    code: `${hAndhxImports}
     const StatelessWithReturn = props => {
  return h(".class");
};`
  },
  {
    title: "Should handle assignment (differs from VariableDeclaration)",
    code: `${hAndhxImports}
     const HandlesAssignment = ({ title }) => {
  title = h("span");
};`
  },
  {
    title: "Should handle h calls in the Array as a parent",
    code: `${hAndhxImports}
     handleArrays = [
  h(Sidebar, { categories }),
  h(CategoryQuestions, { ...question, isBusiness })
];`
  },

  {
    title: "Should handle single class name with dashes (Revolut only)",
    code: `${hAndhxImports}
   const ClassNameWithDashesSingle = props => h(".this-is-dashes");`
  },
  {
    title: "Should handle multiple class names with dashes (Revolut only)",
    code: `${hAndhxImports}
     const ClassNameWithDashesMulti = props => h(".this-is-dashes.dash-afterDash");`
  },
  {
    title: "Should transpile for object property field",
    code: `${hAndhxImports}
     const JustPropField = h(Stuff, {
  children: h(FormattedMessage, { ...commonMessages.learnMore })
});`
  },
  {
    title: "Should transpile for for regular function",
    code: `${hAndhxImports}
     function HyperscriptAsRegularFunction(props) {
  return h("h1");
}`
  },
  {
    title: "Should transpile for VariableDeclaration",
    code: `${hAndhxImports}
     const HyperscriptAsVariable = h("div.lol", {
  someProp: "lol"
});`
  },
  {
    title: "Should wrap nested expressions into {} also transpiling them down if posible",
    code: `${hAndhxImports}
     const HyperscriptWithExpressionAsChildren = h(
  AnotherComponent,
  { foo: "bar", bar: () => ({}), shouldRender: thing.length > 0 },
  [arr.map(() => h("h1"))]
);`
  },
  {
    title: "Should ignore transformation TemplateStrings to StringLiteral if computed expressions found",
    code: `${hAndhxImports}
     // Should be ignored from transforming
const FirstArgTemplateLiteralWithComputedExpressions = h(\`div.lol\${stuff}\`, {
  someProp: "lol"
});`
  },
  {
    title: "Should transform TemplateStrings to StringLiteral if no computed expressions found",
    code: `${hAndhxImports}
const FirstArgTemplateLiteral = h(\`div.lol\`, {
  someProp: "lol"
});`
  },
  {
    title: "Should ignore when first argument is called function",
    code: `${hAndhxImports}
const WhenFirstArgumentIsFunctionThatIsCalled = () =>
  h(getLoadableAnimation("pageCareersDeliver"), [h(fn())]);`
  },
  {
    title:
      "Should ignore computed root call, but transpile nested calls, handle third array argument of different types in proper way",
    code: `${hAndhxImports}
 const ComputedRootWithObjectPropertyDeclaration = () =>
  h(
    ANIMATIONS[country],
    {
      className: "lol",
      content: h(".selectItem", [
        h(".flag", [
          h(RoundFlag, {
            mix: "flag",
            size: "xs",
            code: currencyData.countryCode
          }),
          // Computed not root should be wrapped in {}
          h(ANIMATIONS[country], { className: "lol" })
        ])
      ])
    },
    // This first children in array will be ignored
    [
      h(ANIMATIONS[country], { className: "lol" }),
      h("h1"),
      kek && mem,
      surreal ? lol : kek,
      t.tabName,
      lol,
      <div />
    ]
  )`
  },
  {
    title: "Computed class name BinaryExpression",
    code: `${hAndhxImports}
   h("div" + "div");`
  },
  {
    title: "Ignore third argument when it's not array",
    code: `${hAndhxImports}
     const ThirdArgOnIgnoredIsNotArray = () =>
  h(
    ANIMATIONS[country],
    {
      className: "lol"
    },
    children
  );`
  },
  {
    title: "Ignore second argument when it's not array",
    code: `${hAndhxImports}
     const SecondArgOnIgnoredIsNotArray = () => h(ANIMATIONS[country], children);`
  },
  {
    title: "One level deep member expression self closing",
    code: `${hAndhxImports}
     const MultiMemberExpressionWithClosingTag = () =>
  h(Pricing.lol, { className });`
  },
  {
    title: "One level deep member expression with closing tag",
    code: `${hAndhxImports}
     const MultiMemberExpressionWithClosingTag = () =>
  h(Pricing.lol, { className }, [h("h1")]);`
  },
  {
    title: "Deep member expression with closing tag",
    code: `${hAndhxImports}
     const MultiMemberExpressionWithClosingTag = () =>
  h(Pricing.lol.kek, { className }, [h("h1")]);`
  },
  {
    title: "Deep member expression self closing",
    code: `${hAndhxImports}
     const MultiMemberExpressionWithClosingTag = () =>
  h(Pricing.lol.kek, { className });`
  },
  {
    title: "Should ignore transformation when ConditionalExpression is a first parameter",
    code: `${hAndhxImports}
     h(isCanada ? doStuff : doAnotherStuff, { someProp: true })
     h('div', isCanada ? someProps : anotherProps)
     h('div', isCanada ? someProps : anotherProps, "SomeChildren")
     `
  },
  {
    title:
      'handle h(Abc, { [kek]: 0, ["norm"]: 1 }) to < Abc {...{ [kek]: 0 }} {...{ ["norm" + lol]: 1 }} norm={1} />',
    code: `${hAndhxImports}
const ComplexComputedAttibutesHandling = () =>
  h(Abc, { [kek]: 0, ["norm" + lol]: 1, ["ok"]: 2 });`
  },
  {
    title: "Handle multi classNames css modules (Rev only)",
    code: `${hAndhxImports}
h(".bar.fuzz.stuff", ["bar fuzz"]);`
  },
  {
    title: "Should process children but ignore computed parent",
    code: `${hAndhxImports}
h(\`calcualted \${stuff}\`, { amazing: "stuff" }, [
  h("h1"),
  h("h2"),
  h("h3"),
  h("div", [h("div")])
]);`
  },
  {
    title: "Should handle shouldRender in root call (Revolut only)",
    code: `${hAndhxImports}
  h('div', { shouldRender: true })`
  },
  {
    title: "Should handle shouldRender in nested call (Revolut only)",
    code: `${hAndhxImports}
  h('div', [h('div', { shouldRender: true })]);
  h('div', { className }, [h('div', { shouldRender: true })]);
  `
  },
  {
    title: "Should handle shouldRender when root call is ignored (Revolut only)",
    code: `${hAndhxImports}
  h('div' + 'div', [h('div', { shouldRender: true })]);
  h('div' + 'div', { className }, [h('div', { shouldRender: true }), h('div')]);
  `
  },
  {
    title: "Should handle shouldRender when object property call is ignored (Revolut only)",
    code: `${hAndhxImports}
  h('div' + 'div', [h('div', { shouldRender: true })]);
  h('div' + 'div', { className, field: h('div', { shouldRender: true }) });
  `
  },
  {
    title: "Should handle shouldRender when JSXAttribute property call is ignored (Revolut only)",
    code: `${hAndhxImports}
  <div attr={h('div', { shouldRender: true })} />;
  h('div' + 'div', { className, field: h('div', { shouldRender: true }) });
  `
  }
];

pluginTester({
  pluginName: "hyperscript-to-jsx",
  plugin: hyperToJsxTransform,
  snapshot: true,
  formatResult: output =>
    prettier.format(output, {
      semi: true,
      singleQuote: true,
      parser: "babylon"
    }),
  tests: [
    ...tests,
    ...tests.map(test => ({
      ...test,
      title: `${test.title} Revolut`,
      pluginOptions: { revolut: true }
    }))
  ]
});

module.exports = {
  check
};
