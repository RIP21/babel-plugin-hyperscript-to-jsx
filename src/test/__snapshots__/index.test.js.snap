// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hyperscript-to-jsx Complex real: Complex real 1`] = `
"
import h from \\"react-hyperscript\\";

const StatelessComponent = props => h(\\"h1\\");

let dropdownCurrencies = currencyCodes.map((c) => {
  let currencyData = getCurrency(c)
  let label = isMobile() ? currencyData.id : \`\${currencyData.id} – \${
  currencyData.title
}\`
  return {
    key: c,
    content: h('.selectItem', [
      h('div', label),
      h('.flag', [
        h(RoundFlag, { mix: 'flag', size: 'xs', code: currencyData.countryCode })
      ])
    ])
  }
})

const StatelessWithReturn = props => {
  return h(\\".class\\", { shouldRender: lol.length > 0 });
};

function named(props) {
  return h(\\"h1\\");
}

class Comp extends React.Component {
  render() {
    return h(\\".categories\\", [
      categories.map(({ key, title, children, url: categoryURL }) =>
        h(\\".category\\", { key }, [
          sectionsImages[key] &&
            h(sectionsImages[key], { className: styles.animation }),
          h(Heading, { level: 2, tag: \\"h2\\", spacing: \\"half\\" }, title),
          h(\\"ul\\", [
            children.map((item, index, arr) => {
              const { url, title } = item;
              if (index > 3) {
                return null;
              }
              if (index === 3 && arr.length > 4) {
                return h(\\"li.viewMore\\", { key: item.key }, [
                  h(FormattedMessage, { ...messages.viewMore }, message =>
                    h(HelpLink, {
                      url: enhanceUrlEmbedded(categoryURL),
                      title: message,
                      scrollToTop: true
                    })
                  )
                ]);
              }
              return h(\\"li\\", { key: item.key }, [
                h(HelpLink, { url: enhanceUrlEmbedded(url), title })
              ]);
            })
          ])
        ])
      ),
      h(\\".helpFeatures\\", [
        h(\\".feature\\", [
          h(Link, { href: \\"https://community.revolut.com\\" }, [
            h(\\"img\\", { src: communityIMG }),
            h(\\".text\\", [
              h(FormattedMessage, { ...messages.communityTitle }, message =>
                h(Heading, { level: 6, tag: \\"h4\\", spacing: \\"half\\" }, message)
              ),
              h(FormattedMessage, { ...messages.communitySubtitle })
            ])
          ])
        ]),
        h(\\".feature\\", [
          h(Link, { href: \\"https://blog.revolut.com\\" }, [
            h(\\"img\\", { src: blogIMG }),
            h(\\".text\\", [
              h(FormattedMessage, { ...messages.blogTitle }, message =>
                h(Heading, { level: 6, tag: \\"h4\\", spacing: \\"half\\" }, message)
              ),
              h(FormattedMessage, { ...messages.blogSubtitle })
            ])
          ])
        ]),
        isBusiness && h(\\".feature\\", [
              h(
                Link,
                {
                  href: isEmbeddedBusiness()
                    ? \\"https://www.revolut.com/business/openapi\\"
                    : \\"/business/openapi\\"
                },
                [
                  h(\\"img\\", { src: apiIMG }),
                  h(\\".text\\", [
                    h(FormattedMessage, { ...messages.apiTitle }, message =>
                      h(
                        Heading,
                        { level: 6, tag: \\"h4\\", spacing: \\"half\\" },
                        message
                      )
                    ),
                    h(FormattedMessage, { ...messages.apiSubtitle })
                  ])
                ]
              )
            ])
        ,
        isBusiness
          ? h(\\".feature\\", [
              h(
                Link,
                {
                  href: isEmbeddedBusiness()
                    ? \\"https://www.revolut.com/business/openapi\\"
                    : \\"/business/openapi\\"
                },
                [
                  h(\\"img\\", { src: apiIMG }),
                  h(\\".text\\", [
                    h(FormattedMessage, { ...messages.apiTitle }, message =>
                      h(
                        Heading,
                        { level: 6, tag: \\"h4\\", spacing: \\"half\\" },
                        message
                      )
                    ),
                    h(FormattedMessage, { ...messages.apiSubtitle })
                  ])
                ]
              )
            ])
          : h(\\".feature\\", [
              h(\\"img\\", { src: chatIMG }),
              h(\\".text\\", [
                h(FormattedMessage, { ...messages.chatTitle }, message =>
                  h(Heading, { level: 6, tag: \\"h4\\", spacing: \\"half\\" }, message)
                ),
                h(FormattedMessage, { ...messages.chatSubtitle })
              ])
            ])
      ])
    ]);
  }
}

const stuff = {
  fn() {
    <div>{label}</div>
    h('div', label)
    return <div>{label}</div>
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import React from 'react';
import h from 'react-hyperscript';

const StatelessComponent = props => <h1 />;

let dropdownCurrencies = currencyCodes.map(c => {
  let currencyData = getCurrency(c);
  let label = isMobile()
    ? currencyData.id
    : \`\${currencyData.id} – \${currencyData.title}\`;
  return {
    key: c,
    content: (
      <div className=\\"selectItem\\">
        <div>{label}</div>
        <div className=\\"flag\\">
          <RoundFlag mix=\\"flag\\" size=\\"xs\\" code={currencyData.countryCode} />
        </div>
      </div>
    )
  };
});

const StatelessWithReturn = props => {
  return <div className=\\"class\\" shouldRender={lol.length > 0} />;
};

function named(props) {
  return <h1 />;
}

class Comp extends React.Component {
  render() {
    return (
      <div className=\\"categories\\">
        <div className=\\"helpFeatures\\">
          <div className=\\"feature\\">
            <Link href=\\"https://community.revolut.com\\">
              <img src={communityIMG} />
              <div className=\\"text\\">
                <FormattedMessage {...messages.communityTitle}>
                  {message => (
                    <Heading level={6} tag=\\"h4\\" spacing=\\"half\\">
                      {message}
                    </Heading>
                  )}
                </FormattedMessage>
                <FormattedMessage {...messages.communitySubtitle} />
              </div>
            </Link>
          </div>
          <div className=\\"feature\\">
            <Link href=\\"https://blog.revolut.com\\">
              <img src={blogIMG} />
              <div className=\\"text\\">
                <FormattedMessage {...messages.blogTitle}>
                  {message => (
                    <Heading level={6} tag=\\"h4\\" spacing=\\"half\\">
                      {message}
                    </Heading>
                  )}
                </FormattedMessage>
                <FormattedMessage {...messages.blogSubtitle} />
              </div>
            </Link>
          </div>
          {isBusiness && (
            <div className=\\"feature\\">
              <Link
                href={
                  isEmbeddedBusiness()
                    ? 'https://www.revolut.com/business/openapi'
                    : '/business/openapi'
                }
              >
                <img src={apiIMG} />
                <div className=\\"text\\">
                  <FormattedMessage {...messages.apiTitle}>
                    {message => (
                      <Heading level={6} tag=\\"h4\\" spacing=\\"half\\">
                        {message}
                      </Heading>
                    )}
                  </FormattedMessage>
                  <FormattedMessage {...messages.apiSubtitle} />
                </div>
              </Link>
            </div>
          )}
          {isBusiness ? (
            <div className=\\"feature\\">
              <Link
                href={
                  isEmbeddedBusiness()
                    ? 'https://www.revolut.com/business/openapi'
                    : '/business/openapi'
                }
              >
                <img src={apiIMG} />
                <div className=\\"text\\">
                  <FormattedMessage {...messages.apiTitle}>
                    {message => (
                      <Heading level={6} tag=\\"h4\\" spacing=\\"half\\">
                        {message}
                      </Heading>
                    )}
                  </FormattedMessage>
                  <FormattedMessage {...messages.apiSubtitle} />
                </div>
              </Link>
            </div>
          ) : (
            <div className=\\"feature\\">
              <img src={chatIMG} />
              <div className=\\"text\\">
                <FormattedMessage {...messages.chatTitle}>
                  {message => (
                    <Heading level={6} tag=\\"h4\\" spacing=\\"half\\">
                      {message}
                    </Heading>
                  )}
                </FormattedMessage>
                <FormattedMessage {...messages.chatSubtitle} />
              </div>
            </div>
          )}
        </div>
      </div>
    );
  }
}

const stuff = {
  fn() {
    <div>{label}</div>;
    <div>{label}</div>;
    return <div>{label}</div>;
  }
};

"
`;

exports[`hyperscript-to-jsx Fake all cases: Fake all cases 1`] = `
"
import h from \\"react-hyperscript\\";

const StatelessComponent = props => h(\\"h1\\");

const StatelessWithReturn = props => {
  return h(\\".class\\");
};

function named(props) {
  return h(\\"h1\\");
}

const lol = h(\\"div.lol\\", {
  someProp: \\"lol\\"
});

const ComputedRootWithObjectPropertyDeclaration = () =>
  h(
    ANIMATIONS[country],
    {
      className: \\"lol\\",
      content: h(\\".selectItem\\", [
        h(\\"div\\", label),
        h(\\".flag\\", [
          h(RoundFlag, {
            mix: \\"flag\\",
            size: \\"xs\\",
            code: currencyData.countryCode
          }),
          h(ANIMATIONS[country], { className: \\"lol\\" })
        ])
      ])
    },
    [h(ANIMATIONS[country], { className: \\"lol\\" }), h(\\"h1\\")]
  );

class Comp extends React.Component {
  render() {
    return h(\\"div.example\\", [
      isStuff && h(\\"h1#heading\\", { ...getProps, ...getKnobs(), stuff: \\"\\" }),
      isStuff
        ? h(\\"h1#heading\\", { ...getProps, ...getKnobs(), stuff: \\"\\" })
        : h(\\"h1#heading\\", \\"heading\\"),
      h(\\"h1#heading\\", { ...getProps, ...getKnobs(), stuff: \\"\\" }),
      h(\\"h1#heading\\", getChildren),
      h(ANIMATIONS[country], {
        className: \\"lol\\"
      }),
      h(\\"h1#heading\\", getChildren(), [h(\\"div\\")]),
      h(\\"div\\", [h(\\"div\\", \\"Some content\\")]),
      h(\\"h1#heading\\", \\"This is hyperscript\\"),
      h(\\"h2\\", \\"creating React.js markup\\"),
      h(
        AnotherComponent,
        { foo: \\"bar\\", bar: () => ({}), shouldRender: thing.length > 0 },
        [
          h(\\"li\\", [h(\\"a\\", { href: \\"http://whatever.com\\" }, \\"One list item\\")]),
          h(\\"li\\", \\"Another list item\\")
        ]
      )
    ]);
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import React from 'react';
import h from 'react-hyperscript';

const StatelessComponent = props => <h1 />;

const StatelessWithReturn = props => {
  return <div className=\\"class\\" />;
};

function named(props) {
  return <h1 />;
}

const lol = <div className=\\"lol\\" someProp=\\"lol\\" />;

const ComputedRootWithObjectPropertyDeclaration = () =>
  h(
    ANIMATIONS[country],
    {
      className: 'lol',
      content: (
        <div className=\\"selectItem\\">
          <div>{label}</div>
          <div className=\\"flag\\">
            <RoundFlag mix=\\"flag\\" size=\\"xs\\" code={currencyData.countryCode} />
            {h(ANIMATIONS[country], { className: 'lol' })}
          </div>
        </div>
      )
    },
    [h(ANIMATIONS[country], { className: 'lol' }), h('h1')]
  );

class Comp extends React.Component {
  render() {
    return (
      <div className=\\"example\\">
        {isStuff && <h1 id=\\"heading\\" {...getProps} {...getKnobs()} stuff=\\"\\" />}
        {isStuff ? (
          <h1 id=\\"heading\\" {...getProps} {...getKnobs()} stuff=\\"\\" />
        ) : (
          <h1 id=\\"heading\\">heading</h1>
        )}
        <h1 id=\\"heading\\" {...getProps} {...getKnobs()} stuff=\\"\\" />
        <h1 id=\\"heading\\">{getChildren}</h1>
        {h(ANIMATIONS[country], {
          className: 'lol'
        })}
        <h1 id=\\"heading\\" {...getChildren()}>
          <div />
        </h1>
        <div>
          <div>Some content</div>
        </div>
        <h1 id=\\"heading\\">This is hyperscript</h1>
        <h2>creating React.js markup</h2>
        <AnotherComponent
          foo=\\"bar\\"
          bar={() => ({})}
          shouldRender={thing.length > 0}
        >
          <li>
            <a href=\\"http://whatever.com\\">One list item</a>
          </li>
          <li>Another list item</li>
        </AnotherComponent>
      </div>
    );
  }
}

"
`;
